lib/eventFactory.js

```diff
@@ -2,9 +2,131 @@
 
 const BaseFactory = require('./baseFactory');
 const Event = require('./event');
+const WorkflowParser = require('screwdriver-workflow-parser');
 
 let instance;
 
+/**
+ * Create a job
+ * @method createJob
+ * @param  {JobFactory} jobFactory  Job factory
+ * @param  {Job}        job         Job object
+ * @param  {String}     jobName     Job name
+ * @return {Job}                    Job object
+ */
+function createJob(jobFactory, { permutations, jobName, pipelineId }) {
+    return jobFactory.create({
+        pipelineId,
+        name: jobName,
+        permutations
+    });
+}
+
+/**
+ * Create builds associated with this event
+ * For example, if startFrom ~commit, then create builds with jobs that have requires: ~commit
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {Number}   [config.prNum]             PR number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               (Optional for backwards compatibility)
+ * @param  {Number}   eventId                    Event id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+    let jobsToStart = [];
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    return pipeline.jobs.then((jobs) => {
+        let nextJobs;
+
+        // When startFrom is ~commit
+        if (startFrom === '~commit') {
+            return pipeline.sync()
+            .then((p) => {
+                nextJobs = WorkflowParser.getNextJobs(p.workflowGraph, { trigger: startFrom });
+                jobsToStart = jobs.filter(j => nextJobs.includes(j.name) && j.state === 'ENABLED');
+
+                return Promise.resolve([]);
```

Comment:
should make `jobsToStart` local and return here.

Reply:
You are already in a promise chain. Is returning `Promise.resolve` necessary?